### ISSUE TITLE
Support zero-dimensional tensors as scalars

### ISSUE THREAD

**Issue by fritzo** [pytorch/pytorch#2839]
Note this is a different issue from tensors that have zero-size in one dimension #2055 .

Numpy and Tensorflow both allow consistent treatment of dimensions down to zero dimensions. However torch inconsistently treats a zero-dimensional tensor as empty, rather than containing a single element. For example:
```py
for dim in range(4):
    print(np.ones([2] * dim).sum())
# 1.0
# 2.0
# 4.0
# 8.0

for dim in range(4):
    print(torch.ones([2] * dim).sum())
# 0.0      <---- inconsistency here
# 2.0
# 4.0
# 8.0
```

**Comment by fmassa** [comment]
This issue is being tracked in https://github.com/pytorch/pytorch/issues/1433 , and the solution that will be implemented follows your proposal, i.e., we will support zero-dim tensors, see the comments from the end of the thread.

**Comment by soumith** [comment]
closing in favor of #1433

**Comment by fritzo** [comment]
Thanks @fmassa, it's great to hear that pytorch will be supporting 0-dim tensors soon!
